go-controller/hybrid-overlay/pkg/controller/master.go

```diff
@@ -0,0 +1,263 @@
+
+	kapi "k8s.io/api/core/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+)
+
+// MasterController is the master hybrid overlay controller
+type MasterController struct {
+	kube      *kube.Kube
+	allocator *allocator.SubnetAllocator
+}
+
+// NewMaster a new master controller that listens for node events
+func NewMaster(clientset kubernetes.Interface) (*MasterController, error) {
+	m := &MasterController{
+		kube:      &kube.Kube{KClient: clientset},
+		allocator: allocator.NewSubnetAllocator(),
+	}
+
+	// Add our hybrid overlay CIDRs to the allocator
+	for _, clusterEntry := range config.HybridOverlay.ClusterSubnets {
+		err := m.allocator.AddNetworkRange(clusterEntry.CIDR.String(), 32-clusterEntry.HostSubnetLength)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Mark existing hostsubnets as already allocated
+	existingNodes, err := m.kube.GetNodes()
+	if err != nil {
+		return nil, fmt.Errorf("Error in initializing/fetching subnets: %v", err)
+	}
+	for _, node := range existingNodes.Items {
+		if houtil.IsWindowsNode(&node) {
+			hostsubnet, ok := node.Annotations[types.HybridOverlayNodeSubnet]
+			if ok {
+				klog.V(5).Infof("marking existing node %s hybrid overlay NodeSubnet %s as allocated", node.Name, hostsubnet)
+				if err := m.allocator.MarkAllocatedNetwork(hostsubnet); err != nil {
+					utilruntime.HandleError(err)
+				}
+			}
+		}
+	}
+
+	return m, nil
+}
+
+// Start is the top level function to run hybrid overlay in master mode
+func (m *MasterController) Start(wf *factory.WatchFactory) error {
+	return houtil.StartNodeWatch(m, wf)
+}
+
+func parseHybridOverlayNodeHostSubnet(node *kapi.Node) (*net.IPNet, error) {
+	sub, ok := node.Annotations[types.HybridOverlayNodeSubnet]
+	if !ok {
+		return nil, nil
+	}
+
+	_, subnet, err := net.ParseCIDR(sub)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing node %s annotation %s value %q: %v",
+			node.Name, types.HybridOverlayNodeSubnet, sub, err)
+	}
+
+	return subnet, nil
+}
+
+func sameCIDR(a, b *net.IPNet) bool {
+	if a == b {
+		return true
+	} else if (a == nil && b != nil) || (a != nil && b == nil) {
+		return false
+	}
+	return a.IP.Equal(b.IP) && bytes.Equal(a.Mask, b.Mask)
+}
+
+// updateNodeAnnotation returns:
+// 1) the annotation name
+// 2) the annotation value (if any)
+// 3) true to add the annotation, false to delete it from the node
+// 4) any error that occurred
+func (m *MasterController) updateNodeAnnotation(node *kapi.Node, annotator kube.Annotator) error {
+	extNodeSubnet, _ := parseHybridOverlayNodeHostSubnet(node)
+	ovnNodeSubnet, _ := ovn.ParseNodeHostSubnet(node)
+
+	if !houtil.IsWindowsNode(node) {
+		// Sync/remove subnet annotations for Linux nodes.
+		// - if there is no OVN HostSubnet annotation, remove the HybridOverlayHostSubnet annotation
+		// - if the OVN HostSubnet annotation is different than the HybridOverlayHostSubnet annotation,
+		//   copy the OVN one to HybridOverlayHostSubnet
```

Comment:
@danwinship so that the annotations are consistent. Windows looks for the same annotation that we look for. We clearly aren't going to annotate the Windows nodes with the regular OVN hostsubnet annotation, and I thought it clearest for cluster admins to have consistent annotations at the API side (if not in the code). Is that OK?

Reply:
You're arbitrarily declaring that it would be confusing to use the existing subnet annotation on Windows nodes, but it's not confusing to use the hybrid subnet annotation on Linux nodes.

All Linux nodes are required to be full ovn-kubernetes nodes, and all Windows nodes are required to be vxlan-reachable nodes, so there would not be any ambiguity of meaning if you just used the existing subnet annotation everywhere. Or alternatively, both the Linux and Windows sides need to treat Windows destinations and Linux destinations differently, so it should not be difficult for either of them to deal with if Linux nodes had only the old annotation and Windows nodes had only the new one.